utils/sockaddr.1
.\" sockaddr - list address resolution.
.\" Copyright (C) 2009-2014 David Sugar <dyfet@gnutelephony.org>
.\" Copyright (C) 2015 Cherokees of Idaho.
.\"
.\" This manual page is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU Lesser General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.\"
.\" This manual page is written especially for Debian GNU/Linux.
.\"
.TH sockaddr "1" "January 2010" "GNU uCommon" "GNU Telephony"
.SH NAME
sockaddr \- list address resolution.
.SH SYNOPSIS
.B sockaddr
.RI [ options ]
.I host
.I service
.br
.SH DESCRIPTION
This command lists out the resolved (lookup) addresses for a given host
and service name combination.  If multiple addresses are returned by
the resolver, each valid address is listed.  Services can be specified by
name or port number.
.SH OPTIONS
.TP
.B \-\-udp
Lookup a udp service.
.TP
.B \-\-help
Outputs help screen for the user.
.SH AUTHOR
.B sockaddr
was written by David Sugar <dyfet@gnutelephony.org>.
.SH "REPORTING BUGS"
Report bugs to bug-commoncpp@gnu.org or bugs@gnutelephony.org.
.SH COPYRIGHT
Copyright \(co 2009-2014 David Sugar, Tycho Softworks.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.




ChangeLog
Changes from 6.3.0 to 6.3.1
- copyright assignment to Cherokees of Idaho
- introduction of ucommon cmake hosted macros
- simplified doxygen doc support
- convergence of automake and cmake builds
- make utils easier for source introspection
- converted for cmake builds on debian
- count for String::check should be size_t
- string bounds checking

Changes from 6.2.3 to 6.3.0
- remove legacy cmodel support
- modernize library usage
- simplify autoconf and cmake
- introduce sysruntime...
- cleanup memory operators....

Changes from 6.2.2 to 6.2.3
- a few socket fixes, mostly windows related...

Changes from 6.2.1 to 6.2.2
- bumped gnutls to 3.0.0 or later...
- lots of bug fixes and general cleanup
- socket endian issues and other things for ccrtp
- templated newp and freep handle management

Changes from 6.2.0 to 6.2.1
- datetime: fix const inconsistency
- fsys: cached error variable must be mutable
- replacment new/delete cannot be declared inline
- fix exception specifier clash for c++11

Changes from 6.1.11 to 6.2.0
- standardized cmake paths and stand-alone usages
- gnutls support in cmake
- unified openssl checks, even works for windows
- improved openssl support, new static crypto option
- pkgconfig usable in win32, other win32 fixes
- fixes for mac build & optional disable of utils
- upticked abi version to 7, should have happened with 10/11?

Changes from 6.1.10 to 6.1.11
- any addr and overloads in commoncpp
- cmake abi fix and other apple specific issues
- stream fixes, sflphone recommended changes, and other updates
- key protocol added
- cleaner build for binding without resolver
- cleanup and fixes for commoncpp peering...
- UDPSocket: use ucommon::Socket::address
- Socket::address: add setAny(sockaddr*), setLoopback(sockaddr*)

Changes from 6.1.9 to 6.1.10
- for next release
- iso cleaness
- Fix memory leak in DateTime ctor
- Fix utf8::count()
- secure::erase() should be more secure

Changes from 6.1.8 to 6.1.9
- ost::Socket: use ucommon::Socket for join(), drop() implementations
- back to native threading model for mingw
- Socket::store: use existing len method
- Socket::address: allow printing to std::ostream
- Socket::address: use in_port_t for port numbers
- Socket::join, ::drop: allow to set interface index for IPv6

Changes from 6.1.7 to 6.1.8
- mingw libstdc++ c++11 support now requires winpthread
- more android related updates

Changes from 6.1.6 to 6.1.7
- lots of android related updates

Changes from 6.1.5 to 6.1.6
- some windows and mingw fixes
- Socket::address::print(): minor fixes
- Socket::address: fix constructor documentation
- Socket::address: add unit tests for print()
- Socket::address: add print() method
- Socket::address: add withPort()
- Socket::address: add constructors from standard addr types
- Socket::address: use existing len() method.

Changes from 6.1.4 to 6.1.5
- key merge on load for 6.1.5

Changes from 6.1.3 to 6.1.4
- lots of namespace usage cleanup
- Socket::address: add unit tests for new features
- Socket::address: add isAny, setAny, isLoopback, setLoopback, any, loopback
- Socket::address: convert port to network byte order
- ost::Socket: add compatibility with ucommon::Socket::address
- we don't want commoncpp configs if no lib made
- Socket::address: add equality operator, getPort, setPort, getSize
- improved cmake and added srpm, todo make targets

Changes from 6.1.2 to 6.1.3
- revised unixaddr buffer overflow issue

Changes from 6.1.1 to 6.1.2
- updated contacts, fsf address, and copyright
- revised cmake build
- added setgroups

from ucommon 6.1.0 to 6.1.1
- module flag support for older commoncpp packages
- fix for commoncpp pc file
- some additional original classes mapped to commoncpp
- default Semaphore constructor now valid special case

from ucommon 6.0.7 to 6.1.0
- small fixes based on notes from Tristan
- exception behavior is changed (fixed) for commoncpp
- keywait utility added

from ucommon 6.0.6 to 6.0.7
- error state in fsys open fixed
- fsys error reset inline added

from ucommon 6.0.5 to 6.0.6
- small cleanup of useless validator assignments
- fix gnutls casting and warnings
- modernized automake support

from ucommon 6.0.4 to 6.0.5
- additional constructors to pass pre-allocated memory

from ucommon 6.0.3 to 6.0.4
- fix for address list comparison issue

from ucommon 6.0.2 to 6.0.3
- fixes for older gnutls releases

from ucommon 6.0.1 to 6.0.2
- digest stream operators
- auto memory erasing keystring and keyrandom template
- common code shared for secure library
- some fixes for gnutls support

from ucommon 6.0.0 to 6.0.1
- fixes for gnutls support
- elimination of gcrypt in gnutls

from ucommon 5.5.1 to 6.0.0
- new version 6 api
- fixed string + concat operation
- made StringPager proper base class by naming conventions
- new device access mode and name checking for fsys
- fbuf refocused on process pipe and device access
- more consistent abi naming conventions
- depreciated methods removed
- access base classes now consistent with protocols
- reorg of datetime classes
- allocation and access fault handling

from ucommon 5.5.0 to 5.5.1
- regex checks also for lib, mingw support
- pedantic cleanup of mingw
- anaselli fix for applog and pipes

from ucommon 5.4.3 to 5.5.0
- mapof and listof added
- semaphore default for commoncpp set to 0
- simplify config, no longer separate stdlib and compat
- resolved issues with XOPEN_SOURCE and broken BSD headers 
- many mingw 32 and 64 bit fixes, build fixes for out of source binary builds

from ucommon 5.4.2 to 5.4.3
- fixes for supporting ancient gcc compilers (at least >=2.96!)
- bug in Conditional gettimeout, stack and queue suggested by Pollak

from ucommon 5.4.1 to 5.4.2
- fixes for older gcc builds and template issues.
- bugfix for ConditionalAccess constructor.
- bugfix for commoncpp address functions.

from ucommon 5.4.0 to 5.4.1
- fixes for windows build.

from ucommon 5.3.0 to 5.4.0
- cleanup for ide use.
- NamedList sort uses system collation.

utils/sockaddr.cpp
// Copyright (C) 2010-2014 David Sugar, Tycho Softworks.
// Copyright (C) 2015 Cherokees of Idaho.
//
// This file is part of GNU uCommon C++.
//
// GNU uCommon C++ is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published
// by the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// GNU uCommon C++ is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with GNU uCommon C++.  If not, see <http://www.gnu.org/licenses/>.

#include <ucommon/ucommon.h>

using namespace ucommon;

static shell::flagopt helpflag('h',"--help",    _TEXT("display this list"));
static shell::flagopt althelp('?', NULL, NULL);
static shell::flagopt udp('u', "--udp", _TEXT("lookup udp service"));

int main(int argc, char **argv)
{
    unsigned type = SOCK_STREAM;
    unsigned port;
    char buffer[256];

    shell::bind("sockaddr");
    shell args(argc, argv);

    if(is(helpflag) || is(althelp)) {
        printf("%s\n", _TEXT("Usage: sockaddr [options] arguments..."));
        printf("%s\n\n", _TEXT("Echo command line arguments"));
        printf("%s\n", _TEXT("Options:"));
        shell::help();
        printf("\n%s\n", _TEXT("Report bugs to dyfet@gnu.org"));
        return 0;
    }

    const char *host = args[0];
    const char *service = args[1];

    if(!host || !service)
        shell::errexit(1, "use: sockaddr [options] host service\n");

    if(udp)
        type = SOCK_DGRAM;

    Socket::address addr(host, service, type);
    linked_pointer<struct sockaddr> ap = addr;

    if(!is(ap))
        shell::errexit(2, "*** sockaddr: %s/%s: %s\n",
            host, service, _TEXT("could not find"));

    while(is(ap)) {
        port = 0;
        switch(ap->sa_family) {
        case AF_INET:
            port = ntohs(ap.in()->sin_port);
            break;
#ifdef  AF_INET6
        case AF_INET6:
            port = ntohs(ap.in6()->sin6_port);
            break;
#endif
        }

        if(port) {
            Socket::query(*ap, buffer, sizeof(buffer));
            printf("%s/%u\n", buffer, port);
        }

        ap.next();
    }

    return 0;
}

